### PR TITLE
add hashtag tabs

### PR DESCRIPTION
**Pull request by connyduck**
- For now, supporting only one hashtag. But datastructure and layout is ready for multiple hashtags. I think we need to use the search endpoint with pagination for multiple hashtags.
- there can be multiple hashtag tabs, as opposed to the other tab types

device-2019-03-23-120346

### CODE REVIEW

File: app/src/main/java/com/keylesspalace/tusky/TabData.kt [hunk 1]
```diff
@@ -29,20 +29,22 @@ const val NOTIFICATIONS = "Notifications"
...
-        DIRECT -> TabData(DIRECT, R.string.title_direct_messages, R.drawable.reblog_direct_dark) { ConversationsFragment.newInstance() }
+        HOME -> TabData(HOME, R.string.title_home, R.drawable.ic_home_24dp, { TimelineFragment.newInstance(TimelineFragment.Kind.HOME) })
+        NOTIFICATIONS -> TabData(NOTIFICATIONS, R.string.title_notifications, R.drawable.ic_notifications_24dp, { NotificationsFragment.newInstance() })
+        LOCAL -> TabData(LOCAL, R.string.title_public_local, R.drawable.ic_local_24dp, { TimelineFragment.newInstance(TimelineFragment.Kind.PUBLIC_LOCAL) })
+        FEDERATED -> TabData(FEDERATED, R.string.title_public_federated, R.drawable.ic_public_24dp, { TimelineFragment.newInstance(TimelineFragment.Kind.PUBLIC_FEDERATED) })
+        DIRECT -> TabData(DIRECT, R.string.title_direct_messages, R.drawable.reblog_direct_dark, { ConversationsFragment.newInstance() })
+        HASHTAG -> TabData(HASHTAG, R.string.hashtag, R.drawable.ic_hashtag, { args -> TimelineFragment.newInstance(TimelineFragment.Kind.TAG, args.getOrNull(0).orEmpty()) }, arguments)
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/TabData.kt]: no idea how to get a reference to the `arguments` member inside the `fragment` function so I made a workaround by passing it into the function later

Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/TabData.kt]: There's only one way I can think of and it is `TabData.() -> Fragment` as declaration of a function.

Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/TabData.kt]: Or passing a whole `TabData` to the function, compiles to the same code but maybe is more straightforward, I'm not sure, maybe it's okay like it is now.

Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/TabData.kt]: I will leave it as it is, its not overly complicated

File: app/src/main/java/com/keylesspalace/tusky/TabPreferenceActivity.kt [hunk 2]
```diff
@@ -134,13 +141,73 @@ class TabPreferenceActivity : BaseActivity(), Injectable, ItemInteractionListene
...
+
+    private fun showEditHashtagDialog(tab: TabData? = null) {
+
+        val editText = AppCompatEditText(this)
+        editText.setText("")
+        editText.append(tab?.arguments?.first().orEmpty())
+
+        val dialog = AlertDialog.Builder(this)
+                .setTitle(R.string.edit_hashtag_title)
+                .setMessage(R.string.edit_hashtag_message)
```
Comment (charlag) [inline app/src/main/java/com/keylesspalace/tusky/TabPreferenceActivity.kt]: Why set it to message when it can be a hint?

**Comment by charlag** [comment]
Oh, and of course I want to bother you with formatting .__.
I've tried to look into running pre-commit hook to format today but we need to know where Studio is installed to run this and it sucks.

File: app/src/main/java/com/keylesspalace/tusky/TabPreferenceActivity.kt [hunk 3]
```diff
@@ -134,13 +141,73 @@ class TabPreferenceActivity : BaseActivity(), Injectable, ItemInteractionListene
...
+
+                        currentTabsAdapter.notifyItemChanged(position)
+                    }
+
+                    updateAvailableTabs()
+                    saveTabs()
+                }
+                .create()
+
+        editText.addTextChangedListener(object: TextWatcher {
```
[the remark this answers is not in the document]
Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/TabPreferenceActivity.kt]: nice!

**Comment by aka-rin** [comment]
- Bug: Tag can be empty.
- Advice: View a tag at different instances, since hashtag is an open API without auth in Mastodon: `https://instance.name/api/v1/timelines/tag/tagname`.

**Comment by kajdo** [comment]
Hey ... was looking for the same feature and found this issue ... seems like the merge request didn't make it into a release (at least not on f-droid or I'm to stupid to find it)

**Comment by connyduck** [comment]
It feature was released with Tusky 12 and has been in every release since @kajdo 
Menu -> Account Preferences -> Tabs -> click the + button

**Comment by kajdo** [comment]
Ohhhh ..  I totaly missed that. Being a multi account client it makes sense to have it in account preferences!

Thx a lot

**Pull request merged**